Add S3 cache store

bentsherman opened the pull request:
Close https://github.com/seqeralabs/nf-tower-cloud/issues/4822

Adds a new cache store backed by S3. The cache entry for each task is saved to the following S3 path:
```
s3://<bucket>/<path>/<session-id>/<task-hash>
```

It can be enabled by setting `NXF_CACHE_STORE` to `s3` and `NXF_CACHE_PATH` to the desired S3 path.

Notes:
- ~~Currently included in the nf-amazon plugin, but could be generalized to any object storage since it uses the Path API. Maybe better to make it part of the nextflow module and call it something like `ObjectCacheStore`.~~
- Consider exposing config options in addition to environment variables.
- Consider using buckets for the task hash (like the work directory) instead of storing all task entries under the same S3 prefix. I don't know if this will actually help though. Usually Nextflow just needs to try a particular key, not iterate through all keys.
- Consider setting the default cache path based on the work directory (e.g. `${workDir}/../cache`

pditommaso commented on modules/nextflow/src/main/groovy/nextflow/cache/PathCacheStore.groovy:
```diff
@@ -0,0 +1,170 @@
+    }
+
+    @Override
+    void writeIndex(HashCode key, boolean cached) {
+        getCachePath(key).bytes = [ cached as byte ] as byte[]
+    }
```
Not understanding this logic, the index file is essentially an append log, that keep the order of the execution of the tasks 

https://github.com/nextflow-io/nextflow/blob/s3-cache-store/modules/nextflow/src/main/groovy/nextflow/cache/DefaultCacheStore.groovy#L134-L137

bentsherman replied:
I think I did this because I wasn't sure if the index file and leveldb were independent, i.e. writing a cache entry to leveldb doesn't affect the index file. If they are independent, I will just save the index file as an S3 object and use it exactly as the default store.

bentsherman replied:
However, since I can't append to an S3 object, I will have to re-upload the entire index file every time a task is written. That should be okay for now, but ideally we would want to avoid that.

The index file iteration is used by the `clean` and `log` commands, but they don't need the order of execution. This is why I think we could discard the index file and instead simply list the directory of cache entries when we need to iterate.

pditommaso replied:
Yes, you can create an output stream with `Files.newOutputStream(Path)`

bentsherman replied:
So that does some kind of streaming upload that continues for the entire pipeline run? Seems like magic but should make the index file easy to manage. I will do that.

bentsherman replied:
I know... I'm just curious how the S3 output stream manages the upload under the hood.

pditommaso reviewed (changes requested):
Using the underlying virtual file system is a good idea. I'm concerned to on  503 slow down error that can be reported when resulting large pipeline because the object storage API will be under pressure. 

Also, it would make sense to move this into its own plugin to be activated to enable this feature

bentsherman commented:
> I'm concerned to on 503 slow down error that can be reported when resulting large pipeline because the object storage API will be under pressure.

Me too. The AWS SDK should be able to handle it with it's built-in adaptive retry, but I don't remember if it's enabled by default.

> Also, it would make sense to move this into its own plugin to be activated to enable this feature.

I made it enabled with an environment variable `NXF_CACHE_STORE='path'` so that it doesn't have to reside in a separate plugin. It still uses the default cache store by default.

bentsherman commented:
I moved the patch cache to a plugin and added the index file to behave like the default.

One problem with the plugin is that there is no way to enable it with the `clean` and `log` commands that use the cache. How do you recommend we instrument those commands?

pditommaso commented:
Likely the only way to enable it in those commands is via the NXF_PLUGINS_DEFAULT, which is not very handy. However, the primary need for this plugin is to use it with Tower, therefore I would not care too much

bentsherman commented:
Okay, then this PR is ready for review. I tested with rnaseq-nf by running and then resuming the pipeline. Looks like there is still a problem with `packing.gradle`, but I'm not sure how to fix it.

pditommaso commented on packing.gradle:
```diff
@@ -21,13 +21,14 @@ dependencies {
+    console   project(':plugins:nf-console')
+    ga4gh     project(':plugins:nf-ga4gh')
+    google    project(':plugins:nf-google')
+    pathcache project(':plugins:nf-path-cache')
```
No really needed adding it here, for example nf-codecommit plugin is not include in this list

pditommaso commented on plugins/nf-path-cache/src/main/nextflow/PathCachePlugin.groovy:
```diff
@@ -0,0 +1,31 @@
+ */
+@Slf4j
+@CompileStatic
+class PathCachePlugin extends BasePlugin {
```
What about calling it CloudCacheXxx

pditommaso commented on docs/plugins.md:
```diff
@@ -13,6 +13,7 @@ Currently the following functionalities are implemented as plugin components and
 - `nf-console`: Implement Nextflow [REPL console](https://www.nextflow.io/blog/2015/introducing-nextflow-console.html).
 - `nf-ga4gh`: Support [GA4GH APIs](https://www.ga4gh.org/).
 - `nf-google`: Support for Google cloud.
+- `nf-path-cache`: Support for path-based Nextflow cache.
```
I'd proposed to call this `nf-cloudcache` or `nf-cloudcachedb`, at the end the primary use case is to implement a cache DB for cloud-based runs

pditommaso commented on validation/google.sh:
```diff
@@ -68,4 +64,18 @@ $NXF_CMD -C ./gls.config run ./test-overwrite.nf
+$NXF_CMD -C ./google.config \
+    run nextflow-io/rnaseq-nf \
+    -with-report \
+    -with-trace $OPTS \
```
@sonatype-lift ignore

pditommaso commented:
Ok, [those](https://github.com/nextflow-io/nextflow/pull/4097/commits/24f52ac141cc95b88d0a52e3034a545abb394179) tests works, but it should be added some checks that's effectively resuming the tasks and using the cloud cache. Logging adding some grep in the log file. @bentsherman can you please give it a try? then it can be merged

bentsherman commented on validation/awsbatch.sh:
```diff
@@ -36,7 +36,18 @@ $NXF_CMD run test-complexpaths.nf -resume -c awsbatch.config
+    -with-trace \
+    -plugins nf-cloudcache \
+    -resume
+[[ `grep -c 'Cached process > ' .nextflow.log` == 4 ]] || false
```
@sonatype-lift ignore

pditommaso replied:
this is ok, tho it does not guarantee that's using the S3 based cache, no?

pditommaso replied:
Cool, let's add the same also to Google batch and Azure

bentsherman commented on validation/awsbatch.sh:
```diff
@@ -36,7 +36,20 @@ $NXF_CMD run test-complexpaths.nf -resume -c awsbatch.config
-    -with-trace
+    -with-trace \
+    -plugins nf-cloudcache
+[[ `grep -c 'Using Nextflow cache factory: nextflow.cache.CloudCacheFactory' .nextflow.log` == 1 ]] || false
```
@sonatype-lift ignoreall

pditommaso commented:
I've removed sonatype-lif, just annoying

abhi18av commented:
This is b-e-a-utiful, thanks guys! 🎉

bentsherman commented:
@pditommaso is there a way to trigger the cloudcache plugin when the `NXF_CLOUDCACHE_PATH` variable is set? Currently you have to enable the plugin explicitly, seems like you shouldn't have to do that since it's a core plugin. I'm not sure how to do this through the priority extensions loader.
